eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/immutable/ImmutableHashBag.java

```diff
@@ -677,4 +679,10 @@ protected Object writeReplace()
     {
         return new ImmutableBagSerializationProxy<>(this);
     }
+
+    @Override
+    public SetIterable<T> asSet()
+    {
+        return UnmodifiableMutableSet.of(this.delegate.asSet());
+    }
```

Comment:
It's because implementing an actual immutable version of this view while the underlying bad is itself mutable is a lot of work in this class.

Since I only needed to return a `SetIterable`, I simply protected the integrity of the bag by wrapping it in an unmodifiable decorator.

This would be different if we wanted to return an `ImmutableSet` here...

Reply:
In some places you do return MutableSet though, right? Should we return SetIterable everywhere?

Reply:
Yes, as I explained in the PR description and in https://github.com/eclipse/eclipse-collections/pull/797#discussion_r379922824, I decided to return either `SetIterable` or `MutableSet`.

But we could decide to be more precise and return also `ImmutableSet`, `ImmutableSortedSet`, MutableSortedSet` and so on when possible. It's not always easy, see for example https://github.com/eclipse/eclipse-collections/pull/797#discussion_r379891931 but it could be done I suppose.

Reply:
In my last push, I've used the most precise interface possible everywhere except for the fact that `MutableSet` all returns unmodifiable ones in practice until the invariants can be fixed